File: 3.x.x/crates-utopia/cdn/src/cache/adapter/cloudflare.rs
//! PHP `Utopia\Cdn\Cache\Adapter\Cloudflare`.

use std::sync::Arc;

use serde_json::{json, Value};

use super::Adapter;
use crate::http::{default_client, send, HttpClient, RequestResult};
use crate::{CdnError, Domain};

/// Cloudflare cache-purge adapter.
pub struct Cloudflare {
    zone_id: String,
    api_token: String,
    api_base: String,
    client: Arc<dyn HttpClient>,
}

impl std::fmt::Debug for Cloudflare {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Cloudflare")
            .field("zone_id", &self.zone_id)
            .field("api_base", &self.api_base)
            .finish_non_exhaustive()
    }
}

impl Cloudflare {
    /// URLs per purge request. PHP `PATHS_PER_PURGE`.
    pub const PATHS_PER_PURGE: usize = 30;
    /// Cache tags per purge request. PHP `KEYS_PER_PURGE`.
    pub const KEYS_PER_PURGE: usize = 30;

    #[must_use]
    pub fn new(zone_id: impl Into<String>, api_token: impl Into<String>) -> Self {
        Self {
            zone_id: zone_id.into(),
            api_token: api_token.into(),
            api_base: "https://api.cloudflare.com/client/v4".into(),
            client: Arc::new(default_client()),
        }
    }

    #[must_use]
    pub fn with_client(mut self, client: Arc<dyn HttpClient>) -> Self {
        self.client = client;
        self
    }

    #[must_use]
    pub fn with_api_base(mut self, api_base: impl Into<String>) -> Self {
        self.api_base = api_base.into();
        self
    }

    fn send_body(&self, body: Value) -> Result<(), CdnError> {
        let result = self.request(
            "POST",
            &format!("/zones/{}/purge_cache", self.zone_id),
            Some(&body),
        );
        if !is_success(&result) {
            return Err(CdnError::runtime(format_error(&result)));
        }
        Ok(())
    }

    fn request(&self, method: &str, path: &str, body: Option<&Value>) -> RequestResult {
        let headers = [
            ("User-Agent", "Utopia CDN Cloudflare Adapter".to_owned()),
            ("Authorization", format!("Bearer {}", self.api_token)),
            ("Content-Type", "application/json".to_owned()),
        ];
        send(
            self.client.as_ref(),
            method,
            &format!("{}{path}", self.api_base),
            &headers,
            body,
        )
    }
}

fn is_success(result: &RequestResult) -> bool {
    result.status >= 200
        && result.status < 300
        && result.response.is_object()
        && result.response.get("success") == Some(&Value::Bool(true))
}

fn format_error(result: &RequestResult) -> String {
    let message = result
        .error
        .clone()
        .or_else(|| {
            result
                .response
                .get("errors")
                .and_then(Value::as_array)
                .and_then(|errors| errors.first())
                .and_then(|error| error.get("message"))
                .and_then(Value::as_str)
                .map(str::to_owned)
        })
        .unwrap_or_else(|| "Unknown purge error".into());
    format!(
        "Cloudflare purge failed with status {}: {message}",
        result.status
    )
}

impl Adapter for Cloudflare {
    fn purge_paths(&self, domain: &str, paths: &[String]) -> Result<(), CdnError> {
        let domain = Domain::validate(domain)?;
        let paths = Domain::validate_paths(paths)?;
        if paths.is_empty() {
            return Ok(());
        }
        for chunk in paths.chunks(Self::PATHS_PER_PURGE) {
            let urls: Vec<String> = chunk
                .iter()
                .map(|path| format!("https://{domain}{path}"))
                .collect();
            self.send_body(json!({ "files": urls }))?;
        }
        Ok(())
    }

    fn purge_domain(&self, domain: &str) -> Result<(), CdnError> {
        self.send_body(json!({ "hosts": [Domain::validate(domain)?] }))
    }

    fn purge_keys(&self, keys: &[String]) -> Result<(), CdnError> {
        if keys.is_empty() {
            return Ok(());
        }
        for chunk in keys.chunks(Self::KEYS_PER_PURGE) {
            self.send_body(json!({ "tags": chunk }))?;
        }
        Ok(())
    }

    fn purge_zone(&self) -> Result<(), CdnError> {
        self.send_body(json!({ "purge_everything": true }))
    }
}

File: src/Cdn/Cache/Adapter/Cloudflare.php
<?php

namespace Utopia\Cdn\Cache\Adapter;

use Psr\Http\Client\ClientExceptionInterface;
use Psr\Http\Client\ClientInterface;
use Utopia\Cdn\Cache\Adapter;
use Utopia\Cdn\Domain;
use Utopia\Client;
use Utopia\Client\Adapter\Curl\Client as CurlAdapter;
use Utopia\Psr7\ContentType;
use Utopia\Psr7\Header;
use Utopia\Psr7\Method;
use Utopia\Psr7\Request\Factory as RequestFactory;

class Cloudflare implements Adapter
{
    /**
     * URLs per purge request. Names the batch size this adapter has always used rather than
     * changing it: Cloudflare documents 100 URLs per request, 500 on Enterprise, so 30 is within
     * every plan and can be raised deliberately.
     */
    public const int PATHS_PER_PURGE = 30;

    /**
     * Cache tags per purge request. Cloudflare documents 100 operations per request for tags on
     * every plan; 30 is what this adapter has always sent.
     */
    public const int KEYS_PER_PURGE = 30;

    private readonly ClientInterface $client;

    public function __construct(
        private readonly string $zoneId,
        private readonly string $apiToken,
        ?ClientInterface $client = null,
        private readonly string $apiBase = 'https://api.cloudflare.com/client/v4',
    ) {
        $this->client = $client ?? new Client(new CurlAdapter());
    }

    public function purgePaths(string $domain, array $paths): void
    {
        $domain = Domain::validate($domain);
        $paths = Domain::validatePaths($paths);

        if ($paths === []) {
            return;
        }

        foreach (array_chunk($paths, self::PATHS_PER_PURGE) as $chunk) {
            $urls = array_map(static fn(string $path): string => 'https://' . $domain . $path, $chunk);
            $this->send(['files' => $urls]);
        }
    }

    /**
     * Purges every cached response served for the hostname, and nothing served for another.
     */
    public function purgeDomain(string $domain): void
    {
        $this->send(['hosts' => [Domain::validate($domain)]]);
    }

    public function purgeKeys(array $keys): void
    {
        if ($keys === []) {
            return;
        }

        // Cache tags only match responses the origin tagged with a Cache-Tag header.
        foreach (array_chunk($keys, self::KEYS_PER_PURGE) as $chunk) {
            $this->send(['tags' => $chunk]);
        }
    }

    /**
     * Purges every cached response in the zone, whatever hostname it was served for.
     */
    public function purgeZone(): void
    {
        $this->send(['purge_everything' => true]);
    }

    /**
     * @param array<string, mixed> $body
     */
    private function send(array $body): void
    {
        $result = $this->request(Method::POST, '/zones/' . $this->zoneId . '/purge_cache', $body);

        if (!$this->isSuccess($result)) {
            throw new \RuntimeException($this->formatError($result));
        }
    }

    /**
     * The status line is the primary signal, as it is for Fastly. Cloudflare's
     * envelope carries its own verdict, and an explicit `success: false` inside
     * a 2xx is still a rejected purge — but a 2xx without the envelope is
     * acceptance, so a body a gateway rewrote or a test double never sent does
     * not fail a purge the status line already confirmed.
     *
     * @param array{statusCode:int,response:array<string, mixed>|string|null,error:string|null} $result
     */
    private function isSuccess(array $result): bool
    {
        if ($result['statusCode'] < 200 || $result['statusCode'] >= 300) {
            return false;
        }

        if (\is_array($result['response']) && \array_key_exists('success', $result['response'])) {
            return $result['response']['success'] === true;
        }

        return true;
    }

    /**
     * @param array{statusCode:int,response:array<string, mixed>|string|null,error:string|null} $result
     */
    private function formatError(array $result): string
    {
        $message = $result['error'] ?? null;

        if (\is_array($result['response'])) {
            $message ??= $result['response']['errors'][0]['message'] ?? null;
        }

        $message ??= 'Unknown purge error';

        return 'Cloudflare purge failed with status ' . $result['statusCode'] . ': ' . $message;
    }

    /**
     * @param array<string, mixed>|null $body
     * @return array{statusCode:int,response:array<string, mixed>|string|null,error:string|null}
     */
    private function request(string $method, string $url, ?array $body = null): array
    {
        $request = new RequestFactory()->json($method, $this->apiBase . $url, $body);
        $request = $request
            ->withHeader(Header::USER_AGENT, 'Utopia CDN Cloudflare Adapter')
            ->withHeader(Header::AUTHORIZATION, 'Bearer ' . $this->apiToken)
            ->withHeader(Header::CONTENT_TYPE, ContentType::JSON);

        try {
            $response = $this->client->sendRequest($request);
        } catch (ClientExceptionInterface $error) {
            return ['statusCode' => 0, 'response' => null, 'error' => $error->getMessage()];
        }

        $contents = (string) $response->getBody();

        try {
            $decoded = json_decode($contents, true, flags: JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            $decoded = $contents;
        }

        return ['statusCode' => $response->getStatusCode(), 'response' => $decoded, 'error' => null];
    }
}
